documentation/design-docs/data-catalog.md

```diff
@@ -0,0 +1,521 @@
+# Design doc: Data catalog
+
+This doc describes the design of a data catalog system for improving our
+data documentation, quality, and orchestration.
+
+## Motivation
+
+From the [Architecture Upgrades 2023 project
+README](https://github.com/orgs/ccao-data/projects/11/views/1?pane=info):
+
+> The Data Department has developed data infrastructure to facilitate its core
+> functions: modeling, reporting, publishing open data, and research. This
+> infrastructure uses the system-of-record (iasWorld) as a "base" dataset, then
+> adds third-party, collected, and generated data.
+>
+> The data mostly flows in a single direction: it is extracted from the
+> system-of-record to S3, cleaned and transformed via Athena, then pushed to
+> modeling, open data, etc. The only outputs that currently go back into the
+> system-of-record are predicted values from the residential and condominium
+> models.
+>
+> This approach is simple and has worked well for us, but is difficult to expand,
+> maintain, and trust. Our data stack is essentially a big one-way funnel that
+> feeds the models; it wasn't designed with more complex applications in mind.
+> As the Data Department grows, so too will the needs from our infrastructure.
+
+This document describes the design of a data catalog system that can resolve the
+problems with our current data infrastructure described in the README, problems
+like:
+
+* Referential integrity issues
+* Schema inconsistency
+* Lack of documentation around existing data
+* Orchestration issues (timing, automating data creation)
+* Complex joins and views
+* Lack of data validation and alerting for iasWorld and third-party data
+
+For a full list of issues we intend to solve with the catalog, see the
+[requirements](#requirements).
+
+## Requirements
+
+Each of the following requirements has two components: A motivation ("why do
+do we want to do this?") plus a list of requirements ("how will we do this?").
+
+### Referential integrity
+
+#### Motivation
+
+From the [project
+README](https://github.com/orgs/ccao-data/projects/11?pane=info):
+
+> As far as we can tell, iasWorld does not have keys within its database. This
```

Comment:
Note that this isn't the normal state of affairs. The system definitely defines PK/FK by default, but I think they were dropped for our implementation due to difficulties migrating the data, though I'm not positive about this.